src/twisted/trial/_dist/test/test_disttrial.py

```diff
@@ -369,146 +389,363 @@ def test_runWithoutTestButWithAnError(self):
+        assert_that(self.successResultOf(c), equal_to(42))
 
-            def succeedingRun(self, case, result):
-                called.append(None)
-                if len(called) == 5:
-                    return fail(RuntimeError("oops"))
-                return succeed(None)
+    def test_iterateWhile(self):
+        """
+        ``iterateWhile`` executes the actions from its factory until the predicate
+        does not match an action result.
+        """
+        actions: List[Deferred[int]] = [Deferred(), Deferred(), Deferred()]
 
-        workers = []
-        fakeReactor = CountingReactorWithSuccess(workers)
-        self.addCleanup(self.reap, workers)
+        def predicate(value):
+            return value != 42
+
+        d: Deferred[int] = Deferred.fromCoroutine(
+            iterateWhile(predicate, list(actions).pop)
+        )
+        # Let the action it is waiting on complete
+        actions.pop().callback(7)
+
+        # It does not match the predicate so it is not done yet.
+        self.assertNoResult(d)
+
+        # Let the action it is waiting on now complete - with the result it
+        # wants.
+        actions.pop().callback(42)
+
+        assert_that(self.successResultOf(d), equal_to(42))
+
+    def test_countingCalls(self):
+        """
+        ``countingCalls`` decorates a function so that it is called with an
+        increasing counter and passes the return value through.
+        """
+
+        @countingCalls
+        def target(n: int) -> int:
+            return n + 1
+
+        for expected in range(1, 10):
+            assert_that(target(), equal_to(expected))
 
-        scheduler, cooperator = self.getFakeSchedulerAndEternalCooperator()
 
-        result = self.runner.run(
-            TestCase(), fakeReactor, cooperate=cooperator.cooperate, untilFailure=True
+class WorkerPoolBroken(Exception):
+    """
+    An exception for ``StartedWorkerPoolBroken`` to fail with to allow tests
+    to exercise exception code paths.
+    """
+
+
+class StartedWorkerPoolBroken:
+    """
+    A broken, started worker pool.  Its workers cannot run actions.  They
+    always raise an exception.
+    """
+
+    async def run(self, workerAction: WorkerAction) -> None:
+        raise WorkerPoolBroken()
+
+    async def join(self) -> None:
+        return None
+
+
+@define
+class BrokenWorkerPool:
+    """
+    A worker pool that has workers with a broken ``run`` method.
+    """
+
+    _config: WorkerPoolConfig
+
+    async def start(
+        self, reactor: interfaces.IReactorProcess
+    ) -> StartedWorkerPoolBroken:
+        return StartedWorkerPoolBroken()
+
+
+class _LocalWorker:
+    async def run(self, case: TestCase, result: TestResult) -> None:
+        TrialSuite([case]).run(result)
+
+
+class WorkerBroken(Exception):
+    """
+    A worker tried to run a test case but the worker is broken.
+    """
+
+
+class _BrokenLocalWorker:
+    async def run(self, case: TestCase, result: TestResult) -> None:
```

Comment:
Minor comment.
Maybe be explicit about what is broken with this worker
```suggestion
class _BrokenLocalWorker:
    """
    L{LocalWorker} that always fails to start. 
    """
    async def run(self, case: TestCase, result: TestResult) -> None:
```